src/SetProtocol.js

```diff
@@ -310,36 +308,52 @@ class SetProtocol {
       let tokenInstance = await ERC20.at(tokenAddress);
       let userBalance = await tokenInstance.balanceOf(userAddress);
 
-      return userBalance;  
+      return userBalance;
     } catch (error) {
       console.log('Error in retrieving user balance', error, tokenAddress, userAddress);
-    }   
+    }
+  }
+
+  /**
+   *  Retrieves the decimals of an ERC20 token
+   */
+  async getDecimals(tokenAddress) {
+    try {
+      let tokenInstance = await ERC20.at(tokenAddress);
+      let decimals = await tokenInstance.decimals();
+
+      return decimals || 18;
+    } catch (error) {
+      console.log('Error in retrieving decimals', error, tokenAddress);
+    }
   }
 
   /**
    *  Given a list of tokens, retrieves the user balance as well as token metadata
    */
   async getUserBalancesForTokens(tokenAddresses, userAddress) {
-    let results = [];
-
     // For each token, get all the token metadata
     async function getUserBalanceAndAddtoResults(tokenAddress) {
       let token = { address: tokenAddress };
-      let tokenPromises = [
-        this.getTokenName(tokenAddress),
-        this.getTokenSymbol(tokenAddress),
-        this.getUserBalance(tokenAddress, userAddress)
-      ];
-      let tokenPromiseResults = await Promise.all(tokenPromises);
-      token.name = tokenPromiseResults[0];
-      token.symbol = tokenPromiseResults[1];
-      token.balance = tokenPromiseResults[2];
-      results.push(token);
+      let tokenInstance = await ERC20.at(tokenAddress);
```

Comment:
The reason we use `Promise.all` is to paralleze the async calls work. If you have one `await` after another, it blocks the execution of the next line.